src/dataflow/src/sink/kafka.rs

```diff
@@ -574,3 +607,133 @@ where
 
     Box::new(KafkaSinkToken { shutdown_flag })
 }
+
+/// Encodes a stream of `(Option<Row>, Option<Row>)` updates using Avro.
+///
+/// This operator will only encode `fuel` number of updates per invocation. If necessary, it will
+/// stash updates and use an [`timely::scheduling::Activator`] to re-schedule future invocations.
+///
+/// Input [`Row`] updates must me compatible with the given [`Encoder`].
+///
+/// Updates that are not beyond the given [`SinkAsOf`] and/or the `gate_ts` will be discarded
+/// without encoding them.
+///
+/// Input updates do not have to be partitioned and/or sorted. This operator will not exchange
+/// data. Updates with lower timestamps will be processed before updates with higher timestamps
+/// if they arrive in order. However, this is not a guarantee, as this operator does not wait
+/// for the frontier to signal completeness. It is an optimization for downstream operators
+/// that behave suboptimal when receiving updates that are too far in the future with respect
+/// to the current frontier. The order of updates that arrive at the same timestamp will not be
+/// changed.
+fn avro_encode_stream<G>(
+    input_stream: &Stream<G, ((Option<Row>, Option<Row>), Timestamp, Diff)>,
+    as_of: SinkAsOf,
+    gate_ts: Option<Timestamp>,
+    encoder: Encoder,
+    key_schema_id: Option<i32>,
+    value_schema_id: i32,
+    fuel: usize,
+    name_prefix: String,
+) -> Stream<G, ((Option<Vec<u8>>, Option<Vec<u8>>), Timestamp, Diff)>
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let name = format!("{}-avro_encode", name_prefix);
+
+    let mut builder = OperatorBuilder::new(name, input_stream.scope());
+    let mut input = builder.new_input(&input_stream, Pipeline);
+    let (mut output, output_stream) = builder.new_output();
+    builder.set_notify(false);
```

Comment:
I'm not certain if this is intended, but it communicates to timely that this operator does not need to be notified about changes in its input frontier. It is a bit odd to have this and also a `FrontieredInputHandle`, which suggests that you are interested in the frontier of the input handle. Easiest fix is to remove this line and then chat about the implications of this call. It's not clearly wrong, in that I don't see an `input.frontier` use, but worth discussing!

Reply:
Oh boy, sorry about that one! It's left over from an in-progress version where I added metrics for tracking the frontier. I'll change it to `InputHandle` but leave in `set_notify(false)` because we're not interested in frontier updates. The actual Kafka Producer operator has this on `true` (the default) because it does care about the frontier.

Reply:
Oh boy, sorry about that one! It's left over from an in-progress version where I added metrics for tracking the frontier. I'll change it to `InputHandle` but leave in `set_notify(false)` because we're not interested in frontier updates. The actual Kafka Producer operator has this on `true` (the default) because it does care about the frontier.